ARCHITECTURE.md

```diff
@@ -0,0 +1,50 @@
+# Correctionlib architecture
+
+This document briefly describes the architecture of correctionlib.
+It is meant to provide a good starting point for new contributors to find their way around the codebase.
+It assumes some familiarity with correctionlib as a user.
+
+## Python modules
+
+- `schemav2` module: Pydantic models for correctionlib's data structures
+	- `CorrectionSet` is a list of `Correction`s
+	- `Correction` represents a single correction. Its `data` attribute is of `Content` type and represents the root node of the computation graph for this correction. Corrections also have a list of inputs as well as one output of type `Variable` (basically a pair of a name and a type, int/float/string) and
+	- `Content` is the type of a node in the computation graph of a `Correction`. It's a `Union` of the various types of corrections available: `Binning`, `MultiBinning`, `Category`, `Formula`, `FormulaRef`, `Transform`, `HashPRNG`, float
+- `highlevel` module: user-facing types
+	- `CorrectionSet` is a list of `Correction`s (same as in `schemav2` but focus is on user API rather than defining the schema/structure of the corrections)
+	- `Correction` and `CompoundCorrection` wrap the corresponding C++ evaluator and expose the `evaluate` method
+- `_core` module: a small module that contains the Python facades for the corresponding C++ types, in `__init__.pyi`.
+	- types are `CorrectionSet`, `Correction`, `CompoundCorrection` and `Variable`
+	- the bindings are declared in `src/python.cc`
+
+## C++ sources
+
+`include/correction.h` and `src/correction.cc` contain the the C++ types that perform the actual computations:
+	- a `Variable` type with a name and a type (string, integer, real)
+	- a `CorrectionSet` builds a list of `Corrections`
+	- the `Correction` type, which builds a compute graph of correction nodes
+	- types for the different types of nodes in a correction's compute graph, e.g. `Binning`, `Formula`, each with its `evaluate` method. They are constructed by deserializing a JSON object. `Formula::Formula`, for example, parses a `TFormula` expression in the JSON and builds the corresponding `FormulaAST`
+
+## Typical call sequence to evaluate a correction
+
+`highlevel.Correction.evaluate`
+   -> `_core.Correction.evalv`
```

Comment:
This doesn't render well in the actual markdown. But it can also be replaced by the image, so maybe that's the way to go. Can put it in `docs/` and embed here.

Reply:
done!

Reply:
looks like this: https://github.com/cms-nanoAOD/correctionlib/blob/55b56de18683ce90d663b602ae8f8cddc290b469/ARCHITECTURE.md